RGW/standalone: refactor rgw_zone.h with configstore

AliMasarweh opened the pull request:
https://tracker.ceph.com/issues/57167

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

cbodley commented on src/rgw/driver/rados/rgw_zone.h:
```diff
@@ -525,47 +525,56 @@ WRITE_CLASS_ENCODER(RGWPeriodConfig)
+  virtual ~RGWRealm();
+
+  void encode(bufferlist& bl) const {
     ENCODE_START(1, 1, bl);
-    RGWSystemMetaObj::encode(bl);
+    encode(id, bl);
+    encode(name, bl);
```
you've correctly copied these from `RGWSystemMetaObj::encode()`, but that has an extra `ENCODE_START(1, 1, bl);` and `ENCODE_FINISH(bl);` wrapping them. so i think we need something like this instead:
```diff
   void encode(bufferlist& bl) const {
     ENCODE_START(1, 1, bl);
-    RGWSystemMetaObj::encode(bl);
+    {
+      // these used to be wrapped by RGWSystemMetaObj::encode(),
+      // so the extra ENCODE_START/ENCODE_FINISH are preserved
+      ENCODE_START(1, 1, bl);
+      encode(id, bl);
+      encode(name, bl);
+      ENCODE_FINISH(bl);
+    }
```
and similar for decode() below and DECODE_START/FINISH below

cbodley commented on src/rgw/driver/rados/rgw_zone.h:
```diff
@@ -525,47 +525,56 @@ WRITE_CLASS_ENCODER(RGWPeriodConfig)
-  const std::string& get_predefined_name(CephContext *cct) const override;
-
-  using RGWSystemMetaObj::read_id; // expose as public for radosgw-admin
+  rgw_pool get_pool(CephContext *cct) const;
+  const std::string get_default_oid(bool old_format = false) const;
+  const std::string& get_names_oid_prefix() const;
+  const std::string& get_info_oid_prefix(bool old_format = false) const;
```
anything relating to pools or object names should also be removed

AliMasarweh replied:
I think that the below functions are still required:
```
rgw_pool get_pool(CephContext *cct) const;
const std::string& get_info_oid_prefix(bool old_format = false) const;
```

cbodley replied:
what still calls them?

AliMasarweh replied:
```
string RGWRealm::get_control_oid() const
{
  return get_info_oid_prefix() + id + ".control";
}
```
and **get_control_oidis** called by `int RGWRealmWatcher::watch_start`,  which also calls **get_pool**
```
int RGWRealmWatcher::watch_start(const DoutPrefixProvider *dpp, const RGWRealm& realm)
{
  // code
  rgw_pool pool(realm.get_pool(cct));
  // some code
  auto oid = realm.get_control_oid(); 
  // code
}
&
int RGWSI_Zone::list_realms(const DoutPrefixProvider *dpp, list<string>& realms)
{
  RGWRealm realm;
  RGWSI_SysObj::Pool syspool = sysobj_svc->get_pool(realm.get_pool(cct));

  return syspool.list_prefixed_objs(dpp, realm_names_oid_prefix, &realms);
}
```

cbodley replied:
thanks. these are rados details that will eventually need to go away

`ConfigStore::list_realm_names()` should be able to replace `RGWSI_Zone::list_realms()`

`RGWRealmWatcher` and `RGWRealmReloader` will probably need more thought. if we need to keep get_pool()/get_control_oid() for now, at least leave comment there like `// TODO: use ConfigStore for watch/notify`

AliMasarweh replied:
Got it!

cbodley commented on src/rgw/driver/rados/rgw_zone.h:
```diff
@@ -525,47 +525,56 @@ WRITE_CLASS_ENCODER(RGWPeriodConfig)
 class RGWPeriod;
 
-class RGWRealm : public RGWSystemMetaObj
+class RGWRealm
```
love it!

it looks like `RGWRealm::set_current_period()` can also be removed in favor of `rgw::realm_set_current_period()`

cbodley commented on src/rgw/driver/rados/rgw_zone.h:
```diff
@@ -525,47 +525,56 @@ WRITE_CLASS_ENCODER(RGWPeriodConfig)
-  virtual ~RGWRealm() override;
+  RGWRealm(const std::string& _id, const std::string& _name = "") : id(_id), name(_name) {}
+  RGWRealm(CephContext *_cct): cct(_cct) {}
+  RGWRealm(const std::string& _name, CephContext *_cct, RGWSI_SysObj *_sysobj_svc): name(_name), cct(_cct){}
```
we shouldn't need this RGWSI_SysObj constructor anymore

AliMasarweh commented on src/rgw/rgw_period_puller.cc:
```diff
@@ -68,9 +70,12 @@ int RGWPeriodPuller::pull(const DoutPrefixProvider *dpp, const std::string& peri
-  int r = period.init(dpp, cct, svc.sysobj, y);
+  period.set_epoch(zero_epoch);
+  auto config_store_type = g_conf().get_val<std::string>("rgw_config_store");
+  auto cfgstore = DriverManager::create_config_store(dpp, config_store_type);
```
I have made some progress with this fix, can you take a look if my approach is correct?

cbodley replied:
yes! still a number of `DriverManager::create_config_store()` calls to convert, but you're definitely on the right track

cbodley commented on src/rgw/services/svc_mdlog.h:
```diff
@@ -75,11 +76,11 @@ class RGWSI_MDLog : public RGWServiceInstance
 
   // traverse all the way back to the beginning of the period history, and
   // return a cursor to the first period in a fully attached history
-  RGWPeriodHistory::Cursor find_oldest_period(const DoutPrefixProvider *dpp, optional_yield y);
+  RGWPeriodHistory::Cursor find_oldest_period(const DoutPrefixProvider *dpp, optional_yield y, rgw::sal::ConfigStore* cfgstore);
```
this class `RGWSI_MDLog` already has a `cfgstore` member variable, why does this member function take it too? same for init_oldest_log_period and pull_period below

cbodley commented on src/rgw/driver/rados/rgw_service.cc:
```diff
@@ -113,22 +114,22 @@ int RGWServices_Def::init(CephContext *cct,
 
   can_shutdown = true;
 
-  int r = finisher->start(y, dpp);
+  int r = finisher->start(y, dpp, cfgstore);
```
this passes `cfgstore` to a bunch of classes that don't need it. it's probably much easier to pass `cfgstore` into the constructors of the classes that actually need it. that appears to be just these two?
```c++
  mdlog = std::make_unique<RGWSI_MDLog>(cct, run_sync);
  zone = std::make_unique<RGWSI_Zone>(cct);
```
that way you don't need to change `RGWServiceInstance` or any of its overrides

cbodley commented on src/rgw/driver/rados/rgw_rest_realm.cc:
```diff
@@ -342,11 +363,7 @@ class RGWOp_Realm_List : public RGWRESTOp {
 
 void RGWOp_Realm_List::execute(optional_yield y)
 {
-  {
-    // read default realm
-    RGWRealm realm(driver->ctx(), static_cast<rgw::sal::RadosStore*>(driver)->svc()->sysobj);
-    [[maybe_unused]] int ret = realm.read_default_id(this, default_id, y);
-  }
+  default_id = s->penv.cfgstore->read_default_realm_id(this, y, default_id);
```
this function returns an integer error code, which we had previously ignored with `[[maybe_unused]] int ret`. don't assign to `default_id` here, which is the output argument of the function

cbodley reviewed (commented):
good progress 👍 `RGWZoneGroup` and `RGWZoneParams` still have several read/write member functions and inherit from `RGWSystemMetaObj` - will we be able to get rid of that entirely?

cbodley commented:
i think the rgw/multisite suite in teuthology will be the "final boss" for this pr, but an easy way to get started with smoke testing is:

> ~/ceph/build $ MON=1 OSD=1 RGW=1 MDS=0 MGR=0 ../src/test/rgw/test-rgw-multisite.sh 2 1

this runs vstart in two different subdirectories and sets up a simple multisite configuration between them (a realm with one zonegroup and two zones). that should give reasonable coverage of the admin commands and rest api

AliMasarweh commented on src/rgw/driver/rados/rgw_rest_realm.cc:
```diff
@@ -102,7 +109,9 @@ void RGWOp_Period_Post::execute(optional_yield y)
   auto cct = driver->ctx();
 
   // initialize the period without reading from rados
-  period.init(this, cct, static_cast<rgw::sal::RadosStore*>(driver)->svc()->sysobj, y, false);
+  auto config_store_type = g_conf().get_val<std::string>("rgw_config_store");
+  auto cfgstore = DriverManager::create_config_store(this, config_store_type);
+  cfgstore->read_period(this, y,driver->get_zone()->get_current_period_id(), std::nullopt, period);
```
Ok I got it now.
So I should remove cct and not read at all here?

cbodley replied:
sounds good. if any leftover member functions still depend on `cct` we can pass it as an argument

AliMasarweh commented on src/rgw/driver/rados/rgw_zone.cc:
```diff
@@ -704,14 +706,12 @@ int commit_period(const DoutPrefixProvider* dpp, optional_yield y,
 {
-  auto zone_svc = static_cast<rgw::sal::RadosStore*>(driver)->svc()->zone; // XXX
-
   ldpp_dout(dpp, 20) << __func__ << " realm " << realm.id
       << " period " << current_period.id << dendl;
   // gateway must be in the master zone to commit
-  if (info.master_zone != zone_svc->get_zone_params().id) {
+  if (driver->get_zone()->get_zonegroup().is_master_zonegroup()) {
     error_stream << "Cannot commit period on zone "
-        << zone_svc->get_zone_params().id << ", it must be sent to "
+        << driver->get_zone()->get_id() << ", it must be sent to "
```
I will leave it as it was for now, and later I will try replacing it with `SiteConfig`

AliMasarweh commented:
vstart is failing on:
```
setting up user testid
2025-04-07T07:17:22.395+0000 7f8164e0fa40  0 zone cannot have an empty id
2025-04-07T07:17:22.395+0000 7f8164e0fa40 -1 failed reading zone info: ret -22 (22) Invalid argument
2025-04-07T07:17:22.395+0000 7f8164e0fa40  0 ERROR: failed to start notify service ((22) Invalid argument
2025-04-07T07:17:22.395+0000 7f8164e0fa40  0 ERROR: failed to init services (ret=(22) Invalid argument)
```

AliMasarweh commented:
Normal vstart now works without running into this issue:
> vstart is failing on:
> 
> ```
> setting up user testid
> 2025-04-07T07:17:22.395+0000 7f8164e0fa40  0 zone cannot have an empty id
> 2025-04-07T07:17:22.395+0000 7f8164e0fa40 -1 failed reading zone info: ret -22 (22) Invalid argument
> 2025-04-07T07:17:22.395+0000 7f8164e0fa40  0 ERROR: failed to start notify service ((22) Invalid argument
> 2025-04-07T07:17:22.395+0000 7f8164e0fa40  0 ERROR: failed to init services (ret=(22) Invalid argument)
> ```

running `MON=1 OSD=1 RGW=1 MDS=0 MGR=0 ../src/test/rgw/test-rgw-multisite.sh 2 1` hits this issue:
```
2025-04-07T14:13:18.284+0000 7f71e1d21a40  0 ERROR: current period 6f777de9-f055-4d58-ba6b-422ee813fb67 does not contain zone id 26e96fb4-c055-49af-851b-c1e87f7bac7c
2025-04-07T14:13:18.296+0000 7f71e1d21a40  0 WARNING: period init failed: (2) No such file or directory ... skipping
2025-04-07T14:13:18.296+0000 7f71e1d21a40  0 ERROR: search_realm_conf() failed: ret=-2
2025-04-07T14:13:18.296+0000 7f71e1d21a40  0 ERROR: failed to start notify service ((2) No such file or directory
2025-04-07T14:13:18.296+0000 7f71e1d21a40  0 ERROR: failed to init services (ret=(2) No such file or directory)


```

cbodley commented on src/rgw/rgw_realm.cc:
```diff
@@ -228,7 +73,7 @@ int RGWRealm::find_zone(const DoutPrefixProvider *dpp,
 
   RGWPeriod period(period_id, epoch);
-  int r = period.init(dpp, cct, sysobj_svc, get_id(), y);
+  int r = cfgstore->read_period(dpp, y, period_id, epoch, period);
```
> WARNING: period init failed

i think this is because we're trying to read an empty `period_id`. before, we were passing the realm id (`RGWRealm::get_id()`) to `RGWPeriod::init()`, where it was loading that realm to learn its `current_period`

i guess this is coming from `RGWSI_Zone::search_realm_with_zone()`, which has already loaded this realm. so we should be able to use the member variable `current_period` directly:
```c++
  RGWPeriod period;
  int r = cfgstore->read_period(dpp, y, current_period, std::nullopt, period);
```

AliMasarweh commented:
Running the multisite tests locally, using these [instructions](https://github.com/ceph/ceph/blob/main/src/test/rgw/test_multi.md).
I hit this error:
```
======================================================================
ERROR: test suite for <module 'test_multi' from '/home/alimasa/ceph/src/test/rgw/test_multi.py'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/usr/local/lib/python3.9/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/usr/local/lib/python3.9/site-packages/nose/suite.py", line 316, in setupContext
    try_run(context, names)
  File "/usr/local/lib/python3.9/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/home/alimasa/ceph/src/test/rgw/test_multi.py", line 426, in setup_module
    init(False)
  File "/home/alimasa/ceph/src/test/rgw/test_multi.py", line 241, in init
    c1.start()
  File "/home/alimasa/ceph/src/test/rgw/test_multi.py", line 76, in start
    bash(cmd, env=env)
  File "/home/alimasa/ceph/src/test/rgw/test_multi.py", line 46, in bash
    assert(process.returncode == 0)
AssertionError: 
-------------------- >> begin captured stdout << ---------------------
WARNING: error reading test config. Path can be set through the RGW_MULTI_TEST_CONF env variable

--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
rgw_multi.tests: DEBUG: running cmd: /home/alimasa/ceph/src/mstart.sh c1 -n
rgw_multi.tests: DEBUG: command returned status=1 stdout=Cluster dest path: /home/alimasa/ceph/build/run/c1
monitors base port: 6820
rgw base port: 8001
hostname folio11
ip 172.21.5.161
port 6820
creating /home/alimasa/ceph/build/run/c1/keyring
/home/alimasa/ceph/build/bin/monmaptool: monmap file /tmp/ceph_monmap.1119608
/home/alimasa/ceph/build/bin/monmaptool: generated fsid df840448-d257-4102-86d2-d8e974991811
setting min_mon_release = tentacle
epoch 0
fsid df840448-d257-4102-86d2-d8e974991811
last_changed 2025-04-09T06:10:39.243606+0000
created 2025-04-09T06:10:39.243606+0000
min_mon_release 20 (tentacle)
election_strategy: 1
0: [v2:172.21.5.161:6820/0,v1:172.21.5.161:6821/0] mon.a
1: [v2:172.21.5.161:6822/0,v1:172.21.5.161:6823/0] mon.b
2: [v2:172.21.5.161:6824/0,v1:172.21.5.161:6825/0] mon.c
/home/alimasa/ceph/build/bin/monmaptool: writing epoch 0 to /tmp/ceph_monmap.1119608 (3 monitors)

--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
```

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
jenkins test make check arm64

AliMasarweh commented:
jenkins test make check

AliMasarweh commented:
running `RGW_MULTI_TEST_CONF=./test_multi.conf nosetests test_multi.py -a '!fails_with_rgw'` is passing:
```
----------------------------------------------------------------------
Ran 60 tests in 2330.366s

OK (SKIP=16)
[alimasa@folio11 rgw]$ 
```

AliMasarweh commented:
teuthology run https://pulpito.ceph.com/alimasa-2025-04-12_18:57:25-rgw:multisite-wip-alimasa-rgw-standalone-zone-test-distro-default-smithi/
`ceph/src/test/rgw/rgw_multi/tests.py` is failing

AliMasarweh commented:
looks like this is the issue that we hit in teuthology:
```
======================================================================
ERROR: test_multi.test_multi_object_delete
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/usr/local/lib/python3.9/site-packages/nose/util.py", line 620, in newfunc
    return func(*arg, **kw)
  File "/home/alimasa/ceph/src/test/rgw/rgw_multi/tests.py", line 928, in test_multi_object_delete
    bucket.delete_keys(objnames)
  File "/home/alimasa/.local/lib/python3.9/site-packages/boto/s3/bucket.py", line 728, in delete_keys
    while delete_keys2(headers):
  File "/home/alimasa/.local/lib/python3.9/site-packages/boto/s3/bucket.py", line 722, in delete_keys2
    xml.sax.parseString(body, h)
  File "/usr/lib64/python3.9/xml/sax/__init__.py", line 48, in parseString
    parser.parse(inpsrc)
  File "/usr/lib64/python3.9/xml/sax/expatreader.py", line 111, in parse
    xmlreader.IncrementalParser.parse(self, source)
  File "/usr/lib64/python3.9/xml/sax/xmlreader.py", line 125, in parse
    self.feed(buffer)
  File "/usr/lib64/python3.9/xml/sax/expatreader.py", line 221, in feed
    self._err_handler.fatalError(exc)
  File "/usr/lib64/python3.9/xml/sax/handler.py", line 38, in fatalError
    raise exception
xml.sax._exceptions.SAXParseException: <unknown>:1:426: mismatched tag
```

cbodley commented on src/rgw/services/svc_zone.cc:
```diff
@@ -134,20 +137,21 @@ int RGWSI_Zone::do_start(optional_yield y, const DoutPrefixProvider *dpp)
+  ret = cfgstore->read_period(dpp, y, current_period->get_id(), current_period->epoch, *current_period);
   if (ret < 0 && ret != -ENOENT) {
     ldpp_dout(dpp, 0) << "failed reading current period info: " << " " << cpp_strerror(-ret) << dendl;
     return ret;
   }
 
-  ret = zone_params->init(dpp, cct, sysobj_svc, y);
+  ret = cfgstore->read_default_zone(dpp, y, realm->get_id(), *zone_params, nullptr);
```
sorry, i can't tell from the error you pasted

cbodley replied:
@AliMasarweh the dashboard team reported a regression here in https://tracker.ceph.com/issues/71291 because of this leftover zone/zonegroup logic. i'm testing https://github.com/ceph/ceph/pull/63266 as a fix

cbodley commented on src/rgw/services/svc_zone.cc:
```diff
@@ -134,25 +138,30 @@ int RGWSI_Zone::do_start(optional_yield y, const DoutPrefixProvider *dpp)
-  if (ret < 0 && ret != -ENOENT) {
-    ldpp_dout(dpp, 0) << "failed reading realm info: ret "<< ret << " " << cpp_strerror(-ret) << dendl;
-    return ret;
+  if (site->get_realm().has_value()) {
+    *realm = site->get_realm().value();
+  } else {
+    ret = rgw::read_realm(dpp, y, cfgstore, realm->get_id(), realm->get_name(), *realm);
```
if `SiteConfig` didn't load a realm, there isn't one. we shouldn't need the `else` block, just leave `realm` empty. same for `current_period` below

AliMasarweh commented:
Teuthology run https://pulpito.ceph.com/alimasa-2025-04-23_06:04:14-rgw:multisite-wip-alimasa-rgw-standalone-zone-test-distro-default-smithi/
Now it runs with 2 zonegroups, but we failing this test:
```
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:======================================================================
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:FAIL: rgw_multi.tests.test_bucket_log_trim_after_delete_bucket_secondary_reshard
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_teuthology_dc15ac4a651ab5968cb4ffaa9ef0ff1a02484ea2/virtualenv/lib/python3.10/site-packages/nose/case.py", line 170, in runTest
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_db7705e40154d8f64196c6311eeecf91c2cc6f9a/qa/../src/test/rgw/rgw_multi/tests.py", line 1897, in test_bucket_log_trim_after_delete_bucket_secondary_reshard
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:    assert check_bucket_instance_metadata(zone.zone, test_bucket.name)
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:AssertionError
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2025-04-23T06:58:03.890 INFO:tasks.rgw_multisite_tests:Ran 45 tests in 1590.376s
2025-04-23T06:58:03.891 INFO:tasks.rgw_multisite_tests:
2025-04-23T06:58:03.891 INFO:tasks.rgw_multisite_tests:FAILED (SKIP=2, failures=1)
2025-04-23T06:58:03.891 ERROR:teuthology.run_tasks:Saw exception from tasks.
```

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
jenkins test make check

AliMasarweh commented:
jenkins test make check arm64

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
teuthology run for the RGW suite https://pulpito.ceph.com/alimasa-2025-04-23_14:01:03-rgw-wip-alimasa-rgw-standalone-zone-test-distro-default-smithi/
64 passed and 3 failed
Not sure if this is a know issue:
`Command failed (s3 tests against rgw) on smithi149 with status 1: "source /home/ubuntu/cephtest/tox-venv/bin/activate && cd /home/ubuntu/cephtest/s3-tests-client.0 && S3TEST_CONF=/home/ubuntu/cephtest/archive/s3-tests.client.0.conf BOTO_CONFIG=/home/ubuntu/cephtest/boto-client.0.cfg REQUESTS_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt tox -- -v -m 'not fails_on_rgw and not test_of_sts and not webidentity_test and not fails_with_subdomain and not sse_s3 and cloud_restore and not bucket_logging'"`
The other 2 failures are known issues

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
jenkins test make check arm64

AliMasarweh commented:
jenkins test make check

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
jenkins test api

cbodley commented on src/rgw/driver/rados/rgw_rest_realm.cc:
```diff
@@ -303,7 +309,8 @@ void RGWOp_Realm_Get::execute(optional_yield y)
   // read realm
   realm.reset(new RGWRealm(id, name));
-  op_ret = realm->init(this, g_ceph_context, static_cast<rgw::sal::RadosStore*>(driver)->svc()->sysobj, y);
+  auto config_store_type = g_conf().get_val<std::string>("rgw_config_store");
```
unused variable?

cbodley commented on src/rgw/radosgw-admin/radosgw-admin.cc:
```diff
@@ -4601,6 +4601,7 @@ int main(int argc, const char **argv)
                                         false,
 					false, // No background tasks!
                                         null_yield,
+          cfgstore.get(),
 					need_cache && g_conf()->rgw_cache_enabled,
```
nit: indentation looks off

cbodley commented on src/rgw/rgw_rest_ratelimit.cc:
```diff
@@ -101,7 +105,9 @@ void RGWOp_Ratelimit_Info::execute(optional_yield y)
     std::string realm_id = driver->get_zone()->get_realm_id();
     RGWPeriodConfig period_config;
-    op_ret = period_config.read(this, static_cast<rgw::sal::RadosStore*>(driver)->svc()->sysobj, realm_id, y);
+    auto config_store_type = g_conf().get_val<std::string>("rgw_config_store");
```
unused variable

cbodley commented on src/rgw/rgw_rest_ratelimit.cc:
```diff
@@ -302,10 +308,13 @@ void RGWOp_Ratelimit_Set::execute(optional_yield y)
     return;
   }
+
+  auto config_store_type = g_conf().get_val<std::string>("rgw_config_store");
```
unused variable

cbodley commented on src/rgw/services/svc_notify.h:
```diff
@@ -30,6 +30,7 @@ class RGWSI_Notify : public RGWServiceInstance
   RGWSI_Zone *zone_svc{nullptr};
   librados::Rados *rados{nullptr};
   RGWSI_Finisher *finisher_svc{nullptr};
+  rgw::sal::ConfigStore *cfgstore{nullptr};
```
unused member variable

cbodley commented on src/rgw/services/svc_sys_obj_cache.h:
```diff
@@ -26,6 +26,7 @@ class RGWSI_SysObj_Cache : public RGWSI_SysObj_Core
   ObjectCache cache;
 
   std::shared_ptr<RGWSI_SysObj_Cache_CB> cb;
+  rgw::sal::ConfigStore *cfgstore{nullptr};
```
unused member variable

AliMasarweh commented on src/rgw/driver/rados/rgw_rest_realm.cc:
```diff
@@ -102,7 +101,7 @@ void RGWOp_Period_Post::execute(optional_yield y)
   auto cct = driver->ctx();
 
   // initialize the period without reading from rados
-  period.init(this, cct, static_cast<rgw::sal::RadosStore*>(driver)->svc()->sysobj, y, false);
+  s->penv.cfgstore->read_period(this, y, driver->get_zone()->get_current_period_id(), std::nullopt, period);
```
so the read is unnecessary and I should remove it

AliMasarweh commented on src/rgw/driver/rados/rgw_rest_realm.cc:
```diff
@@ -142,7 +144,11 @@ void RGWOp_Period_Post::execute(optional_yield y)
   // if period id is empty, handle as 'period commit'
   if (period.get_id().empty()) {
-    op_ret = period.commit(this, driver, realm, current_period, error_stream, y);
+    if (op_ret < 0) {
+      cerr << "Error initializing realm: " << cpp_strerror(-op_ret) << std::endl;
+      return;
+    }
```
yes, I guess it was pasted by there a mistake and didn't pay attention

AliMasarweh commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -482,6 +483,7 @@ class RGWSyncProcessorThread : public RGWRadosThread {
 class RGWMetaSyncProcessorThread : public RGWSyncProcessorThread
 {
   RGWMetaSyncStatusManager sync;
+  rgw::sal::ConfigStore* cfgstore;
```
Your are correct, they are not initialized anywhere
I will fix that

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
jenkins test api

AliMasarweh commented:
after rebasing of main and finalizing the addresses issues
passing teuthology run https://pulpito.ceph.com/alimasa-2025-04-30_04:17:24-rgw-wip-alimasa-rgw-standalone-zone-test-distro-default-smithi/

cbodley reviewed (approved):
ship it! 🚀

AliMasarweh commented:
> thanks @AliMasarweh, great work!

You are welcome, thanks for you help!
